File: python/airforce/resources/account.py
"""Account self-service, API keys, and billing resources."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from ._base import AsyncAPIResource, SyncAPIResource, clean, enc

_SESSION = "session"
_APIKEY = "api_key"


class Account(SyncAPIResource):
    def me(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/me", auth=_SESSION, **kw)

    def usage(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/usage", auth=_SESSION, **kw)

    def my_usage(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/my-usage", auth=_SESSION, **kw)

    def update(self, *, username: Optional[str] = None, email: Optional[str] = None, password: Optional[str] = None, **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/update", auth=_SESSION, json=clean(username=username, email=email, password=password), **kw)

    def request_password_reset(self, *, email: str, locale: Optional[str] = None, **kw: Any) -> Any:
        return self._transport.request("POST", "/api/auth/request-password-reset", auth="none", json=clean(email=email, locale=locale), **kw)

    def reset_password(self, *, token: str, new_password: str, **kw: Any) -> Any:
        return self._transport.request("POST", "/api/auth/reset-password", auth="none", json={"token": token, "new_password": new_password}, **kw)

    def referral_code(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/referral/code", auth=_SESSION, **kw)

    def referred_users(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/referral/referred-users", auth=_SESSION, **kw)

    def get_price_caps(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/price-caps", auth=_SESSION, **kw)

    def set_price_caps(self, caps: Dict[str, Any], **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/price-caps", auth=_SESSION, json={"caps": caps}, **kw)

    def delete_price_cap(self, model: str, **kw: Any) -> Any:
        return self._transport.request("DELETE", f"/api/user/price-caps/{enc(model)}", auth=_SESSION, **kw)

    def get_model_aliases(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/model-aliases", auth=_SESSION, **kw)

    def set_model_alias(self, *, alias: str, model: str, **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/model-aliases", auth=_SESSION, json={"alias": alias, "model": model}, **kw)

    def set_model_aliases_batch(self, aliases: List[Dict[str, str]], **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/model-aliases/batch", auth=_SESSION, json=aliases, **kw)

    def delete_model_alias(self, alias: str, **kw: Any) -> Any:
        return self._transport.request("DELETE", f"/api/user/model-aliases/{enc(alias)}", auth=_SESSION, **kw)

    def get_model_defaults(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/model-defaults", auth=_SESSION, **kw)

    def set_model_default(self, model: str, **default: Any) -> Any:
        return self._transport.request("PUT", f"/api/user/model-defaults/{enc(model)}", auth=_SESSION, json=clean(**default))

    def delete_model_default(self, model: str, **kw: Any) -> Any:
        return self._transport.request("DELETE", f"/api/user/model-defaults/{enc(model)}", auth=_SESSION, **kw)

    def get_smart_routing(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/smart-routing", auth=_APIKEY, **kw)

    def set_smart_routing(self, groups: Dict[str, Any], **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/smart-routing", auth=_APIKEY, json={"groups": groups}, **kw)

    def test_smart_routing(self, model: str, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/smart-routing/test", auth=_APIKEY, params={"model": model}, **kw)

    def get_channel_prefs(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/channel-prefs", auth=_APIKEY, **kw)

    def set_channel_pins(self, pins: Dict[str, str], **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/channel-prefs", auth=_APIKEY, json=pins, **kw)

    def set_routing_category_prefs(self, prefs: Dict[str, str], **kw: Any) -> Any:
        """Pin models to routing categories (``{model: category_id}``)."""
        return self._transport.request("PUT", "/api/user/routing-category-prefs", auth=_APIKEY, json=prefs, **kw)

    def set_channel_order_prefs(self, prefs: Dict[str, Any], **kw: Any) -> Any:
        """Per-model channel ordering (``{model: {"order": [...], "auto_fallback"?}}``)."""
        return self._transport.request("PUT", "/api/user/channel-order-prefs", auth=_APIKEY, json=prefs, **kw)

    def get_custom_categories(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/custom-categories", auth=_APIKEY, **kw)

    def set_custom_categories(self, categories: List[Dict[str, Any]], **kw: Any) -> Any:
        """Replace the caller's custom routing categories (max 20)."""
        return self._transport.request("PUT", "/api/user/custom-categories", auth=_APIKEY, json=categories, **kw)

    def routing_categories(self, model: str, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/user/routing-categories", auth=_APIKEY, params={"model": model}, **kw)

    def create_custom_model(self, *, fake_name: str, endpoint: str, **params: Any) -> Any:
        """Register a custom provider model (``{fake_name, endpoint, api_key?, ...}``,
        SSRF-validated server-side)."""
        return self._transport.request("POST", "/api/models", auth=_SESSION, json={"fake_name": fake_name, "endpoint": endpoint, **clean(**params)})

    def update_custom_model(self, fake_name: str, **params: Any) -> Any:
        return self._transport.request("PUT", f"/api/models/{enc(fake_name)}", auth=_SESSION, json=clean(**params))

    def delete_custom_model(self, fake_name: str, **kw: Any) -> Any:
        return self._transport.request("DELETE", f"/api/models/{enc(fake_name)}", auth=_SESSION, **kw)

    def sessions(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/me/sessions", auth=_SESSION, **kw)

    def revoke_session(self, jti: str, **kw: Any) -> Any:
        return self._transport.request("DELETE", f"/api/me/sessions/{enc(jti)}", auth=_SESSION, **kw)

    def revoke_other_sessions(self, **kw: Any) -> Any:
        return self._transport.request("DELETE", "/api/me/sessions", auth=_SESSION, **kw)

    def login_history(self, *, limit: Optional[int] = None, **kw: Any) -> Any:
        return self._transport.request("GET", "/api/me/login-history", auth=_SESSION, params=clean(limit=limit), **kw)

    def reset_api_key(self, **kw: Any) -> Any:
        return self._transport.request("POST", "/api/user/reset-api-key", auth=_SESSION, **kw)

    def set_primary_allowed_ips(self, allowed_ips: List[str], **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/primary-allowed-ips", auth=_SESSION, json={"allowed_ips": allowed_ips}, **kw)

    def set_backup_pool_enabled(self, enabled: bool, **kw: Any) -> Any:
        return self._transport.request("PUT", "/api/user/backup-pool-enabled", auth=_APIKEY, json={"enabled": enabled}, **kw)

    def toggle_pay_as_you_go(self, **kw: Any) -> Any:
        return self._transport.request("POST", "/api/pay-as-you-go/toggle", auth=_SESSION, **kw)

    def delete_account(self, *, password: str, totp_code: Optional[str] = None, forfeit_balance_ack: Optional[bool] = None, **kw: Any) -> Any:
        """Soft-close the account (re-authenticates in the body; ``totp_code`` is
        required when 2FA is enrolled). Revokes all sessions and OAuth tokens,
        rotates the primary key, disables secondary keys, and cancels
        subscriptions. The remaining balance is forfeited only when
        ``forfeit_balance_ack=True``. Idempotent — a repeat call returns
        ``{"closed": true}``."""
        return self._transport.request(
            "DELETE", "/api/me/account", auth=_SESSION,
            json=clean(password=password, totp_code=totp_code, forfeit_balance_ack=forfeit_balance_ack), **kw)

    def reactivate(self, *, email: str, password: str, **kw: Any) -> Any:
        """Reactivate a soft-closed account within the 14-day grace window, using
        the account's former email + password. Public (no session) and rate-limited
        like login; no session is minted — log in normally afterwards."""
        return self._transport.request("POST", "/auth/reactivate", auth="none", json={"email": email, "password": password}, **kw)


class AsyncAccount(AsyncAPIResource):
    async def me(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/me", auth=_SESSION, **kw)

    async def usage(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/usage", auth=_SESSION, **kw)

    async def my_usage(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/my-usage", auth=_SESSION, **kw)

    async def update(self, *, username: Optional[str] = None, email: Optional[str] = None, password: Optional[str] = None, **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/update", auth=_SESSION, json=clean(username=username, email=email, password=password), **kw)

    async def request_password_reset(self, *, email: str, locale: Optional[str] = None, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/auth/request-password-reset", auth="none", json=clean(email=email, locale=locale), **kw)

    async def reset_password(self, *, token: str, new_password: str, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/auth/reset-password", auth="none", json={"token": token, "new_password": new_password}, **kw)

    async def referral_code(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/referral/code", auth=_SESSION, **kw)

    async def referred_users(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/referral/referred-users", auth=_SESSION, **kw)

    async def get_price_caps(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/price-caps", auth=_SESSION, **kw)

    async def set_price_caps(self, caps: Dict[str, Any], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/price-caps", auth=_SESSION, json={"caps": caps}, **kw)

    async def delete_price_cap(self, model: str, **kw: Any) -> Any:
        return await self._transport.request("DELETE", f"/api/user/price-caps/{enc(model)}", auth=_SESSION, **kw)

    async def get_model_aliases(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/model-aliases", auth=_SESSION, **kw)

    async def set_model_alias(self, *, alias: str, model: str, **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/model-aliases", auth=_SESSION, json={"alias": alias, "model": model}, **kw)

    async def set_model_aliases_batch(self, aliases: List[Dict[str, str]], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/model-aliases/batch", auth=_SESSION, json=aliases, **kw)

    async def delete_model_alias(self, alias: str, **kw: Any) -> Any:
        return await self._transport.request("DELETE", f"/api/user/model-aliases/{enc(alias)}", auth=_SESSION, **kw)

    async def get_model_defaults(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/model-defaults", auth=_SESSION, **kw)

    async def set_model_default(self, model: str, **default: Any) -> Any:
        return await self._transport.request("PUT", f"/api/user/model-defaults/{enc(model)}", auth=_SESSION, json=clean(**default))

    async def delete_model_default(self, model: str, **kw: Any) -> Any:
        return await self._transport.request("DELETE", f"/api/user/model-defaults/{enc(model)}", auth=_SESSION, **kw)

    async def get_smart_routing(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/smart-routing", auth=_APIKEY, **kw)

    async def set_smart_routing(self, groups: Dict[str, Any], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/smart-routing", auth=_APIKEY, json={"groups": groups}, **kw)

    async def test_smart_routing(self, model: str, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/smart-routing/test", auth=_APIKEY, params={"model": model}, **kw)

    async def get_channel_prefs(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/channel-prefs", auth=_APIKEY, **kw)

    async def set_channel_pins(self, pins: Dict[str, str], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/channel-prefs", auth=_APIKEY, json=pins, **kw)

    async def set_routing_category_prefs(self, prefs: Dict[str, str], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/routing-category-prefs", auth=_APIKEY, json=prefs, **kw)

    async def set_channel_order_prefs(self, prefs: Dict[str, Any], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/channel-order-prefs", auth=_APIKEY, json=prefs, **kw)

    async def get_custom_categories(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/custom-categories", auth=_APIKEY, **kw)

    async def set_custom_categories(self, categories: List[Dict[str, Any]], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/custom-categories", auth=_APIKEY, json=categories, **kw)

    async def routing_categories(self, model: str, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/user/routing-categories", auth=_APIKEY, params={"model": model}, **kw)

    async def create_custom_model(self, *, fake_name: str, endpoint: str, **params: Any) -> Any:
        return await self._transport.request("POST", "/api/models", auth=_SESSION, json={"fake_name": fake_name, "endpoint": endpoint, **clean(**params)})

    async def update_custom_model(self, fake_name: str, **params: Any) -> Any:
        return await self._transport.request("PUT", f"/api/models/{enc(fake_name)}", auth=_SESSION, json=clean(**params))

    async def delete_custom_model(self, fake_name: str, **kw: Any) -> Any:
        return await self._transport.request("DELETE", f"/api/models/{enc(fake_name)}", auth=_SESSION, **kw)

    async def sessions(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/me/sessions", auth=_SESSION, **kw)

    async def revoke_session(self, jti: str, **kw: Any) -> Any:
        return await self._transport.request("DELETE", f"/api/me/sessions/{enc(jti)}", auth=_SESSION, **kw)

    async def revoke_other_sessions(self, **kw: Any) -> Any:
        return await self._transport.request("DELETE", "/api/me/sessions", auth=_SESSION, **kw)

    async def login_history(self, *, limit: Optional[int] = None, **kw: Any) -> Any:
        return await self._transport.request("GET", "/api/me/login-history", auth=_SESSION, params=clean(limit=limit), **kw)

    async def reset_api_key(self, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/user/reset-api-key", auth=_SESSION, **kw)

    async def set_primary_allowed_ips(self, allowed_ips: List[str], **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/primary-allowed-ips", auth=_SESSION, json={"allowed_ips": allowed_ips}, **kw)

    async def set_backup_pool_enabled(self, enabled: bool, **kw: Any) -> Any:
        return await self._transport.request("PUT", "/api/user/backup-pool-enabled", auth=_APIKEY, json={"enabled": enabled}, **kw)

    async def toggle_pay_as_you_go(self, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/pay-as-you-go/toggle", auth=_SESSION, **kw)

    async def delete_account(self, *, password: str, totp_code: Optional[str] = None, forfeit_balance_ack: Optional[bool] = None, **kw: Any) -> Any:
        return await self._transport.request(
            "DELETE", "/api/me/account", auth=_SESSION,
            json=clean(password=password, totp_code=totp_code, forfeit_balance_ack=forfeit_balance_ack), **kw)

    async def reactivate(self, *, email: str, password: str, **kw: Any) -> Any:
        return await self._transport.request("POST", "/auth/reactivate", auth="none", json={"email": email, "password": password}, **kw)


class Keys(SyncAPIResource):
    def create(self, **params: Any) -> Any:
        return self._transport.request("POST", "/v1/keys", json=clean(**params))

    def list(self, **kw: Any) -> List[Any]:
        res = self._transport.request("GET", "/v1/keys", **kw)
        return res.get("keys", []) if isinstance(res, dict) else res

    def update(self, key: str, **params: Any) -> Any:
        return self._transport.request("PATCH", f"/v1/keys/{enc(key)}", json=clean(**params))

    def delete(self, key: str, **kw: Any) -> Any:
        return self._transport.request("DELETE", f"/v1/keys/{enc(key)}", **kw)


class AsyncKeys(AsyncAPIResource):
    async def create(self, **params: Any) -> Any:
        return await self._transport.request("POST", "/v1/keys", json=clean(**params))

    async def list(self, **kw: Any) -> List[Any]:
        res = await self._transport.request("GET", "/v1/keys", **kw)
        return res.get("keys", []) if isinstance(res, dict) else res

    async def update(self, key: str, **params: Any) -> Any:
        return await self._transport.request("PATCH", f"/v1/keys/{enc(key)}", json=clean(**params))

    async def delete(self, key: str, **kw: Any) -> Any:
        return await self._transport.request("DELETE", f"/v1/keys/{enc(key)}", **kw)


class Billing(SyncAPIResource):
    def create_checkout(self, *, plan: str, **params: Any) -> Any:
        return self._transport.request("POST", "/api/creem/create-checkout", auth=_SESSION, json={"plan": plan, **clean(**params)})

    def create_crypto_invoice(self, *, plan: str, **params: Any) -> Any:
        return self._transport.request("POST", "/api/create-nowpayments-invoice", auth=_SESSION, json={"plan": plan, **clean(**params)})

    def create_portal_session(self, **kw: Any) -> Any:
        return self._transport.request("POST", "/api/create-portal-session", auth=_SESSION, json={}, **kw)

    def analytics(self, **kw: Any) -> Any:
        return self._transport.request("GET", "/v1/analytics", auth="none", **kw)


class AsyncBilling(AsyncAPIResource):
    async def create_checkout(self, *, plan: str, **params: Any) -> Any:
        return await self._transport.request("POST", "/api/creem/create-checkout", auth=_SESSION, json={"plan": plan, **clean(**params)})

    async def create_crypto_invoice(self, *, plan: str, **params: Any) -> Any:
        return await self._transport.request("POST", "/api/create-nowpayments-invoice", auth=_SESSION, json={"plan": plan, **clean(**params)})

    async def create_portal_session(self, **kw: Any) -> Any:
        return await self._transport.request("POST", "/api/create-portal-session", auth=_SESSION, json={}, **kw)

    async def analytics(self, **kw: Any) -> Any:
        return await self._transport.request("GET", "/v1/analytics", auth="none", **kw)

File: go/account.go
package airforce

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

// AccountService accesses /api/me and /api/user/*.
type AccountService struct{ client *Client }

// Me returns the current user profile + quotas.
func (s *AccountService) Me(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/me", "session", nil, &out)
	return out, err
}

// Usage returns the account + billing summary.
func (s *AccountService) Usage(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/usage", "session", nil, &out)
	return out, err
}

// MyUsage returns the detailed usage log + aggregates.
func (s *AccountService) MyUsage(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/my-usage", "session", nil, &out)
	return out, err
}

// Update changes username / email / password.
func (s *AccountService) Update(ctx context.Context, params map[string]any) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/update", "session", params, &out)
	return out, err
}

// RequestPasswordReset sends a reset email.
func (s *AccountService) RequestPasswordReset(ctx context.Context, email, locale string) (map[string]any, error) {
	body := map[string]any{"email": email}
	if locale != "" {
		body["locale"] = locale
	}
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/auth/request-password-reset", "none", body, &out)
	return out, err
}

// ResetPassword completes a password reset.
func (s *AccountService) ResetPassword(ctx context.Context, token, newPassword string) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/auth/reset-password", "none", map[string]any{"token": token, "new_password": newPassword}, &out)
	return out, err
}

// ReferralCode returns the caller's referral code.
func (s *AccountService) ReferralCode(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/referral/code", "session", nil, &out)
	return out, err
}

// ReferredUsers lists users referred by the caller.
func (s *AccountService) ReferredUsers(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/referral/referred-users", "session", nil, &out)
	return out, err
}

// GetPriceCaps returns per-model price ceilings.
func (s *AccountService) GetPriceCaps(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/user/price-caps", "session", nil, &out)
	return out, err
}

// SetPriceCaps replaces all per-model price caps.
func (s *AccountService) SetPriceCaps(ctx context.Context, caps map[string]any) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/price-caps", "session", map[string]any{"caps": caps}, &out)
	return out, err
}

// DeletePriceCap removes one model's price cap.
func (s *AccountService) DeletePriceCap(ctx context.Context, model string) error {
	return s.client.deleteJSON(ctx, "/api/user/price-caps/"+url.PathEscape(model), "session", nil)
}

// GetModelAliases returns all model aliases.
func (s *AccountService) GetModelAliases(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/user/model-aliases", "session", nil, &out)
	return out, err
}

// SetModelAlias adds or updates one alias.
func (s *AccountService) SetModelAlias(ctx context.Context, alias, model string) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/model-aliases", "session", map[string]any{"alias": alias, "model": model}, &out)
	return out, err
}

// SetModelAliasesBatch upserts multiple aliases.
func (s *AccountService) SetModelAliasesBatch(ctx context.Context, aliases []map[string]string) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/model-aliases/batch", "session", aliases, &out)
	return out, err
}

// DeleteModelAlias removes one alias.
func (s *AccountService) DeleteModelAlias(ctx context.Context, alias string) error {
	return s.client.deleteJSON(ctx, "/api/user/model-aliases/"+url.PathEscape(alias), "session", nil)
}

// GetModelDefaults returns all per-model parameter defaults.
func (s *AccountService) GetModelDefaults(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/user/model-defaults", "session", nil, &out)
	return out, err
}

// SetModelDefault sets per-model parameter defaults.
func (s *AccountService) SetModelDefault(ctx context.Context, model string, def map[string]any) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/model-defaults/"+url.PathEscape(model), "session", def, &out)
	return out, err
}

// DeleteModelDefault clears a model's defaults.
func (s *AccountService) DeleteModelDefault(ctx context.Context, model string) error {
	return s.client.deleteJSON(ctx, "/api/user/model-defaults/"+url.PathEscape(model), "session", nil)
}

// GetSmartRouting returns the user's smart-routing config.
func (s *AccountService) GetSmartRouting(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/user/smart-routing", "api_key", nil, &out)
	return out, err
}

// SetSmartRouting sets the user's smart-routing groups.
func (s *AccountService) SetSmartRouting(ctx context.Context, groups map[string]any) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/smart-routing", "api_key", map[string]any{"groups": groups}, &out)
	return out, err
}

// TestSmartRouting resolves a model through smart routing.
func (s *AccountService) TestSmartRouting(ctx context.Context, model string) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/user/smart-routing/test", "api_key", url.Values{"model": {model}}, &out)
	return out, err
}

// GetChannelPrefs returns per-model channel preferences.
func (s *AccountService) GetChannelPrefs(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/user/channel-prefs", "api_key", nil, &out)
	return out, err
}

// SetChannelPins sets per-model channel pins.
func (s *AccountService) SetChannelPins(ctx context.Context, pins map[string]string) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/channel-prefs", "api_key", pins, &out)
	return out, err
}

// Sessions lists active sessions.
func (s *AccountService) Sessions(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/me/sessions", "session", nil, &out)
	return out, err
}

// RevokeSession revokes one session by jti.
func (s *AccountService) RevokeSession(ctx context.Context, jti string) error {
	return s.client.deleteJSON(ctx, "/api/me/sessions/"+url.PathEscape(jti), "session", nil)
}

// RevokeOtherSessions revokes all sessions except the current one.
func (s *AccountService) RevokeOtherSessions(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.deleteJSON(ctx, "/api/me/sessions", "session", &out)
	return out, err
}

// LoginHistory returns the caller's login audit trail.
func (s *AccountService) LoginHistory(ctx context.Context, limit int) (map[string]any, error) {
	var query url.Values
	if limit > 0 {
		query = url.Values{"limit": {strconv.Itoa(limit)}}
	}
	var out map[string]any
	err := s.client.getJSON(ctx, "/api/me/login-history", "session", query, &out)
	return out, err
}

// ResetAPIKey rotates the primary API key.
func (s *AccountService) ResetAPIKey(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/user/reset-api-key", "session", nil, &out)
	return out, err
}

// SetPrimaryAllowedIPs sets the IP whitelist for the primary key.
func (s *AccountService) SetPrimaryAllowedIPs(ctx context.Context, ips []string) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/primary-allowed-ips", "session", map[string]any{"allowed_ips": ips}, &out)
	return out, err
}

// SetBackupPoolEnabled toggles the backup pool.
func (s *AccountService) SetBackupPoolEnabled(ctx context.Context, enabled bool) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSONMethod(ctx, http.MethodPut, "/api/user/backup-pool-enabled", "api_key", map[string]any{"enabled": enabled}, &out)
	return out, err
}

// TogglePayAsYouGo flips pay-as-you-go spending.
func (s *AccountService) TogglePayAsYouGo(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/pay-as-you-go/toggle", "session", nil, &out)
	return out, err
}

// --- keys --------------------------------------------------------------------

// APIKey is a provisioned API key.
type APIKey struct {
	Key             string   `json:"key"`
	Label           string   `json:"label,omitempty"`
	Disabled        bool     `json:"disabled,omitempty"`
	Tier            string   `json:"tier"`
	RPMLimit        *int     `json:"rpm_limit,omitempty"`
	CreditAllowance *float64 `json:"credit_allowance,omitempty"`
	LimitReset      string   `json:"limit_reset,omitempty"`
	AllowedModels   []string `json:"allowed_models"`
	AllowedIPs      []string `json:"allowed_ips"`
}

// KeysService accesses /v1/keys.
type KeysService struct{ client *Client }

// Create makes a secondary key. The full key is returned only here.
func (s *KeysService) Create(ctx context.Context, params map[string]any) (*APIKey, error) {
	var out APIKey
	err := s.client.postJSON(ctx, "/v1/keys", "api_key", params, &out)
	return &out, err
}

// List returns the secondary keys (masked).
func (s *KeysService) List(ctx context.Context) ([]APIKey, error) {
	var out struct {
		Keys []APIKey `json:"keys"`
	}
	err := s.client.getJSON(ctx, "/v1/keys", "api_key", nil, &out)
	return out.Keys, err
}

// Update changes a secondary key's settings.
func (s *KeysService) Update(ctx context.Context, key string, params map[string]any) (*APIKey, error) {
	var out APIKey
	err := s.client.postJSONMethod(ctx, http.MethodPatch, "/v1/keys/"+url.PathEscape(key), "api_key", params, &out)
	return &out, err
}

// Delete removes a secondary key.
func (s *KeysService) Delete(ctx context.Context, key string) error {
	return s.client.deleteJSON(ctx, "/v1/keys/"+url.PathEscape(key), "api_key", nil)
}

// --- billing -----------------------------------------------------------------

// BillingService accesses billing/plans/analytics.
type BillingService struct{ client *Client }

// CreateCheckout creates a Creem checkout session.
func (s *BillingService) CreateCheckout(ctx context.Context, params map[string]any) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/creem/create-checkout", "session", params, &out)
	return out, err
}

// CreateCryptoInvoice creates a NowPayments crypto invoice.
func (s *BillingService) CreateCryptoInvoice(ctx context.Context, params map[string]any) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/create-nowpayments-invoice", "session", params, &out)
	return out, err
}

// CreatePortalSession creates a Creem customer-portal session.
func (s *BillingService) CreatePortalSession(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.postJSON(ctx, "/api/create-portal-session", "session", map[string]any{}, &out)
	return out, err
}

// Analytics returns the public global analytics snapshot.
func (s *BillingService) Analytics(ctx context.Context) (map[string]any, error) {
	var out map[string]any
	err := s.client.getJSON(ctx, "/v1/analytics", "none", nil, &out)
	return out, err
}
